Endpoint-Specific Middleware

lewisjkl opened the pull request:
Adds the ability to provide a middleware that can provide HttpApp[F] => HttpApp[F] conversions based on the data contained in the service and endpoint (such as Hints). This will allow more idiomatic implementations of things such as authentication/authorization and other mechanisms that rely on inspecting traits on a per-endpoint basis.

More docs to come as part of the PR.

lewisjkl commented:
@Baccata if you are happy with the contents of this PR, then I will move forward with adding tests in the project and docs for the microsite. Want to make sure the interface is solid before I move on to that though. Let me know!

lewisjkl commented on modules/http4s/src-ce3/Compat.scala:
```diff
@@ -18,7 +18,7 @@ package smithy4s.http4s
   trait Package {
-    private[smithy4s] type EffectCompat[F[_]] = cats.effect.Concurrent[F]
-    private[smithy4s] val EffectCompat = cats.effect.Concurrent
+    private[smithy4s] type EffectCompat[F[_]] = cats.effect.Async[F]
+    private[smithy4s] val EffectCompat = cats.effect.Async
```
Apparently needed for http4s client.fromHttpApp.

lewisjkl commented:
Ah yeah... `cats.effect.kernel.Unique` wasn't a thing until CE3 right? I am going to need to investigate how we may be able to implement this with the older version of http4s + CE2. Maybe just some compat classes that handle the differences.. hopefully.

kubukoz commented on modules/http4s/src/smithy4s/http4s/SmithyHttp4sRouter.scala:
```diff
@@ -23,23 +23,38 @@ import cats.implicits._
 )(implicit effect: EffectCompat[F]) {
 
+  private val pathParamsKey = {
+    implicit val cheatUnique: Unique[Id] = new Unique[Id] {
```
lol

kubukoz commented on modules/http4s/src/smithy4s/http4s/EndpointSpecificMiddleware.scala:
```diff
@@ -0,0 +1,38 @@
+object EndpointSpecificMiddleware {
+
+  trait Simple[F[_]] extends EndpointSpecificMiddleware[F] {
+    def prepareUsingHints(
```
```suggestion
    def prepareWithHints(
```

Baccata replied:
Why not

kubukoz commented on modules/http4s/src/smithy4s/http4s/EndpointSpecificMiddleware.scala:
```diff
@@ -0,0 +1,38 @@
+// format: off
+trait EndpointSpecificMiddleware[F[_]] {
+  def prepare[Alg[_[_, _, _, _, _]]](service: Service[Alg])(
+      endpoint: Endpoint[service.Operation, _, _, _, _, _]
```
I like this so much.

also, is there an argument for why Endpoint couldn't have IEOSESO as type members?

lewisjkl replied:
Hmm I'm not sure it'd be possible because the same middleware implementation will be used across multiple endpoints in a service so the IOESISO will change depending on the endpoint

kubukoz replied:
no I mean they'd be type members _in_ `Endpoint`. I think it might just be an inconvenience because it makes `Transformation[Endpoint[Op, ...]]` impossible to express directly

lewisjkl replied:
Oh I see, sorry I misread your comment. Yeah I'm not sure about that one tbh

Baccata replied:
It may be achievable, but it should be its own PR. Want to have a go at it @kubukoz ?

kubukoz replied:
perhaps

kubukoz reviewed (changes requested):
This is really cool! I have a couple nitpicks as always.

kubukoz commented on modules/http4s/src/smithy4s/http4s/SmithyHttp4sRouter.scala:
```diff
@@ -23,23 +23,38 @@ import cats.implicits._
+      def unique: Id[Unique.Token] = new Unique.Token()
+
+    }
+    Key.newKey[Id, smithy4s.http.PathParams]
```
http4s creates keys this way

```scala
val ConnectionInfo: Key[Connection] = Key.newKey[SyncIO, Connection].unsafeRunSync()
```

it's `SyncIO`, so it's fine to `.unsafeRunSync()` on any platform. I'd say let's do the same

Baccata replied:
Meh, that's an abuse imho. My point of view is that http4s should not depend on cats-effect-core, but cats-effect-std (or even -kernel). I wish there was just an `unsafeCreate` on `Key`.

kubukoz replied:
yeah, it should not, but it does ;)

(I think it's because fs2 has an explicit SyncIO compiler)

> I wish there was just an `unsafeCreate` on Key.

you know what to do...

Baccata replied:
> you know what to do... 

I do : tell @lewisjkl to ignore your PR comment and keep the current implementation that is totally fine  :trollface:

kubukoz replied:
Just know that the `new Unique` gives me the heebiejeebies :P

Baccata replied:
Well, it may well be that we need to use `SyncIO` in CE2 :/

daddykotex commented on modules/guides/src/smithy4s/guides/Auth.scala:
```diff
@@ -0,0 +1,131 @@
+        }
+        .getOrElse(IO.pure(false))
+
+      isAuthorized.flatMap {
```
I'm not sure which one I prefer, but there is `isAuthorized.ifM(inputApp(request), IO.raiseError(new NotAuthorizedError("Not authorized!")))` that would work here

kubukoz replied:
I would just recommend that if you use `ifM`, you always use named arguments

kubukoz replied:
alternatively you can do  `isAuthorized.ensure(new NotAuthorizedError(...))(identity) *> inputApp(request)` or something like that

lewisjkl replied:
Yeah I was trying to stay lighter on the cats syntax since this is a walkthrough targeted at maybe more beginner users, but I guess maybe being more idiomatic is more important.

kubukoz replied:
at least you didn't do: `isAuthorized.flatMap(if(_) ... else ...)` :)

lewisjkl commented:
Alright, I addressed comments, added tests, and added a guide to the docs. Give me your feedback on what needs to improve/change/etc so we can get this merged 🚀

kubukoz commented on modules/http4s/test/src/smithy4s/http4s/EndpointSpecificMiddlewareSpec.scala:
```diff
@@ -0,0 +1,160 @@
+    ): HttpApp[IO] => HttpApp[IO] = { inputApp =>
+      HttpApp[IO] { request =>
+        val hasTag: (Hints, String) => Boolean = (hints, tagName) =>
+          hints.get[smithy.api.Tags].exists(_.value.contains(tagName))
```
first time I'm seeing anyone use this hint for anything in Scala

lewisjkl replied:
Hah yeah it was just a really easy way I could think to test that the hints were being sent through properly. I don't have any real use cases for it atm. Though I guess it could come in handy for some arbitrary middleware thing 🤷

kubukoz commented:
amazing stuff :)

Baccata commented on modules/http4s/src/smithy4s/http4s/internals/SmithyHttp4sClientEndpoint.scala:
```diff
@@ -79,12 +82,16 @@ private[http4s] class SmithyHttp4sClientEndpointImpl[F[_], Op[_, _, _, _, _], I,
 // format: on
 
+  private val transformedClient: Client[F] =
+    Client.fromHttpApp[F](middleware(client.toHttpApp))
```
Right, so unfortunately, I don't think we should rely on `fromHttpApp` in non-test code, because it's a [non-trivial transformation](https://github.com/http4s/http4s/blob/series/0.23/client/shared/src/main/scala/org/http4s/client/Client.scala#L278) and I'm somewhat scared that we'd be giving a footgun to the users, as it's easy to mess up the request lifecycle in client-side middleware. 

Essentially the main problem is that the HttpApp abstraction should work against `Resource[F, *]` instead of `F`. I believe there's an issue somewhere in http4s, but the impact will be quite heavy on the ecosystem so I think they're waiting for further stabilisation before pulling the trigger on that. 

So, we need to require the middleware to be provided as a `Client[F] => Client[F]`, which unfortunately implies creating another construct to echo the `EndpointSpecificMiddleware`, which may mean that we'd have to rename that middleware to `EndpointSpecificServerMiddleware which is a bit of a mouthful. 

How about `ServerEndpointMiddleware` and `ClientEndpointMiddleware` ?

lewisjkl replied:
Added in https://github.com/disneystreaming/smithy4s/pull/614/commits/75313f5a407bb6021c304029c64b9a6439e14b0f
